### PR TITLE
Initial prep work for streaming tools more conveniently

### PR DESCRIPTION
**Pull request by willbakst**
In preparation for adding convenience around streaming tools, I've added:
- `Partial` class from Pydantic issues page around dyanmically recreating a partial class (like `Partial` in TypeScript)
- Ability to stream the raw `tool_calls` more easily using our `OpenAITool` class.

### CODE REVIEW

File: tests/test_partial.py [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+class PartialDeeperModel(BaseModel):
+    """A deeper model."""
+
+    shallow: Optional[PartialShallowModel] = {}  # type: ignore
+    param: Optional[str] = None
+
+
+def test_deeper_partial():
+    """Tests that `Partial` works to make all fields optional in a deeper model."""
+    print(Partial[DeeperModel].model_json_schema())
```
Comment (brenkao) [inline tests/test_partial.py:47]: remove prints

File: mirascope/partial.py
Comment (brenkao) [inline mirascope/partial.py:1]: does this make sense to be called typing, so people use from mirascope.typing import Partial. similar to from typing import Optional

File: tests/chat/models/test_async_openai_chat.py [hunk 2]
```diff
@@ -166,6 +166,43 @@ async def test_async_openai_chat_stream_messages_kwarg(
...
+    "prompt,tools",
+    [
+        ("fixture_foobar_prompt", ["fixture_my_tool"]),
+        ("fixture_foobar_prompt", ["fixture_my_tool", "fixture_empty_tool"]),
+    ],
+)
+async def test_async_openai_chat_stream_tools(
+    mock_create, fixture_chat_completion_chunk_with_tools, prompt, tools, request
+):
+    """Tests that `OpenAIChat` returns the expected response when called with tools."""
```
Comment (brenkao) [inline tests/chat/models/test_async_openai_chat.py:183]: AsyncOpenAIChat

File: mirascope/chat/utils.py [hunk 3]
```diff
@@ -12,26 +12,45 @@
...
-        return _get_openai_chat_messages(prompt)
+        return [
+            cast(ChatCompletionMessageParam, {"role": role, "content": content})
+            for role, content in prompt.messages
+        ]
     else:
         return [cast(ChatCompletionMessageParam, {"role": "user", "content": prompt})]
 
 
+def setup_openai_kwargs(
```
Comment (brenkao) [inline mirascope/chat/utils.py:28]: not sure if there is a protocol for this but, perhaps patch_openai_kwargs or something to that effect informing us that you are mutating the dictionary.

File: mirascope/chat/utils.py [hunk 4]
```diff
@@ -12,26 +12,45 @@
...
+    if tools:
+        openai_tools: list[type[OpenAITool]] = [
+            tool if isclass(tool) else convert_function_to_openai_tool(tool)
+            for tool in tools
+        ]
+        kwargs["tools"] = [tool.tool_schema() for tool in openai_tools]
+        if "tool_choice" not in kwargs:
+            kwargs["tool_choice"] = "auto"
+
+        return openai_tools
```
Comment (brenkao) [inline mirascope/chat/utils.py:49]: we currently arent doing anything with this return so I would omit it completely. that way this function is very clear in what it is doing which is mutating kwargs, whereas currently it feels like its mutating kwargs but returning something slightly different

File: mirascope/chat/utils.py (hunk 3 above)
Comment (willbakst) [reply mirascope/chat/utils.py]: Done.

File: mirascope/partial.py
Comment (willbakst) [reply mirascope/partial.py:1]: Hmm, we can always add a typing module if we want, but I wanted to have a standalone reference source file to the original author here. And they can import `from mirascope import Partial` in the meantime

File: tests/chat/models/test_async_openai_chat.py (hunk 2 above)
Comment (willbakst) [reply tests/chat/models/test_async_openai_chat.py:183]: Done.

File: tests/test_partial.py (hunk 1 above)
Comment (willbakst) [reply tests/test_partial.py:47]: Done.

**Review by brenkao — APPROVED** [review]

**Pull request merged**